File: src/memento/backup.py
from __future__ import annotations

import hashlib
import json
import shutil
import sqlite3
import subprocess
import tarfile
from dataclasses import dataclass
from pathlib import Path
from tempfile import TemporaryDirectory
from typing import Any

from memento.app import MementoRuntime, runtime_paths_for
from memento.config import ServiceConfig
from memento.repository.git import materialize_current_checkout


@dataclass(frozen=True, slots=True)
class BackupManifest:
    schema_version: int
    repo_revision: str
    files: dict[str, str]

    def as_dict(self) -> dict[str, Any]:
        return {
            "schema_version": self.schema_version,
            "repo_revision": self.repo_revision,
            "files": self.files,
        }


MANIFEST_NAME = "manifest.json"


def create_backup(runtime: MementoRuntime, destination: Path) -> BackupManifest:
    destination.mkdir(parents=True, exist_ok=True)
    repo_tar = destination / "repo.git.tar.gz"
    control_copy = destination / "control.sqlite"
    derived_copy = destination / "derived.sqlite"

    with tarfile.open(repo_tar, "w:gz") as archive:
        archive.add(runtime.paths.repo_paths.bare_dir, arcname="repo.git")
    _sqlite_backup(runtime.paths.control_db, control_copy)
    if runtime.paths.derived_db.exists():
        _sqlite_backup(runtime.paths.derived_db, derived_copy)

    files = {
        repo_tar.name: _sha256(repo_tar),
        control_copy.name: _sha256(control_copy),
    }
    if derived_copy.exists():
        files[derived_copy.name] = _sha256(derived_copy)

    manifest = BackupManifest(
        schema_version=1,
        repo_revision=runtime.status_snapshot()["repo_revision"],
        files=files,
    )
    (destination / MANIFEST_NAME).write_text(
        json.dumps(manifest.as_dict(), indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )
    return manifest


def restore_backup(
    config: ServiceConfig,
    backup_dir: Path,
    *,
    rebuild_derived: bool = True,
) -> dict[str, Any]:
    manifest_path = backup_dir / MANIFEST_NAME
    manifest = BackupManifest(**json.loads(manifest_path.read_text(encoding="utf-8")))
    for name, digest in manifest.files.items():
        path = backup_dir / name
        if not path.exists():
            raise ValueError(f"backup file is missing: {name}")
        if _sha256(path) != digest:
            raise ValueError(f"checksum mismatch for {name}")

    paths = runtime_paths_for(config)
    temp_parent = paths.root.parent if paths.root.parent.exists() else backup_dir.parent
    with TemporaryDirectory(dir=str(temp_parent)) as temp_dir:
        staging_root = Path(temp_dir) / "state"
        staging_root.mkdir(parents=True, exist_ok=True)
        with tarfile.open(backup_dir / "repo.git.tar.gz", "r:gz") as archive:
            _safe_extract_tar(archive, staging_root)
        shutil.copy2(backup_dir / "control.sqlite", staging_root / "control.sqlite")
        derived_backup = backup_dir / "derived.sqlite"
        derived_verified = manifest.files.get("derived.sqlite")
        if not rebuild_derived and derived_verified is not None:
            shutil.copy2(derived_backup, staging_root / "derived.sqlite")
        staged_config = config.model_copy(
            update={
                "repository": config.repository.model_copy(update={"root_path": str(staging_root)})
            }
        )
        staged_paths = runtime_paths_for(staged_config)
        archived_head = subprocess.run(
            [
                "git",
                "--git-dir",
                str(staged_paths.repo_paths.bare_dir),
                "rev-parse",
                "refs/heads/main",
            ],
            capture_output=True,
            text=True,
            check=True,
        ).stdout.strip()
        if archived_head != manifest.repo_revision:
            raise ValueError("backup manifest revision does not match archived main")
        materialize_current_checkout(staged_paths.repo_paths, revision=manifest.repo_revision)
        if rebuild_derived:
            from memento.derived.index import DerivedIndex

            derived = DerivedIndex(staged_paths.derived_db)
            derived.rebuild(
                staged_paths.repo_paths.current_dir,
                repo_revision=manifest.repo_revision,
            )
        backup_old = paths.root.with_name(paths.root.name + ".pre-restore")
        if backup_old.exists():
            shutil.rmtree(backup_old)
        if paths.root.exists():
            paths.root.rename(backup_old)
        staging_root.rename(paths.root)
        if backup_old.exists():
            shutil.rmtree(backup_old)
    return {
        "repo_revision": manifest.repo_revision,
        "restored_root": str(paths.root),
        "rebuild_derived": rebuild_derived,
    }


def _sqlite_backup(source_path: Path, target_path: Path) -> None:
    source = sqlite3.connect(source_path)
    target = sqlite3.connect(target_path)
    try:
        source.backup(target)
    finally:
        target.close()
        source.close()


def _safe_extract_tar(archive: tarfile.TarFile, destination: Path) -> None:
    destination = destination.resolve()
    for member in archive.getmembers():
        if member.issym() or member.islnk():
            raise ValueError(f"refusing to restore linked archive member: {member.name}")
        target = (destination / member.name).resolve()
        if destination not in target.parents and target != destination:
            raise ValueError(
                f"refusing to restore archive member outside destination: {member.name}"
            )
    archive.extractall(destination, filter="data")


def _sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        while True:
            block = handle.read(65_536)
            if not block:
                break
            digest.update(block)
    return digest.hexdigest()

File: go/service/backup.go
package service

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/rcarmo/memento/go/derived"
	"github.com/rcarmo/memento/go/repository"
	"io"
	"io/fs"
	_ "modernc.org/sqlite"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const backupManifestName = "manifest.json"

type BackupManifest struct {
	SchemaVersion int               `json:"schema_version"`
	RepoRevision  string            `json:"repo_revision"`
	Files         map[string]string `json:"files"`
}

func rejectNestedStatePath(path, root, kind string) error {
	candidate, _ := filepath.Abs(path)
	state, _ := filepath.Abs(root)
	relative, _ := filepath.Rel(state, candidate)
	if relative == "." || relative != ".." && !strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
		return fmt.Errorf("%s must be outside repository root_path", kind)
	}
	return nil
}
func sha256File(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	digest := sha256.New()
	if _, err = io.CopyBuffer(digest, file, make([]byte, 65536)); err != nil {
		return "", err
	}
	return hex.EncodeToString(digest.Sum(nil)), nil
}
func sqliteSnapshot(ctx context.Context, source, target string) error {
	db, _ := sql.Open("sqlite", source)
	defer db.Close()
	escaped := strings.ReplaceAll(target, "'", "''")
	_, err := db.ExecContext(ctx, "VACUUM INTO '"+escaped+"'")
	return err
}
func archiveDirectory(source, target, name string) error {
	return archiveDirectoryWith(source, target, name, os.Create, filepath.WalkDir, os.Open, tar.FileInfoHeader)
}
func archiveDirectoryWith(source, target, name string, create func(string) (*os.File, error), walk func(string, fs.WalkDirFunc) error, open func(string) (*os.File, error), headerFor func(fs.FileInfo, string) (*tar.Header, error)) error {
	file, err := create(target)
	if err != nil {
		return err
	}
	gz := gzip.NewWriter(file)
	gz.Header.ModTime = time.Unix(0, 0)
	archive := tar.NewWriter(gz)
	walkErr := walk(source, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		relative, _ := filepath.Rel(source, path)
		archivePath := name
		if relative != "." {
			archivePath = filepath.ToSlash(filepath.Join(name, relative))
		}
		info, err := entry.Info()
		if err != nil {
			return err
		}
		header, err := headerFor(info, "")
		if err != nil {
			return err
		}
		header.Name = archivePath
		header.ModTime = time.Unix(0, 0)
		header.AccessTime = time.Time{}
		header.ChangeTime = time.Time{}
		header.Uid = 0
		header.Gid = 0
		header.Uname = ""
		header.Gname = ""
		if err = archive.WriteHeader(header); err != nil {
			return err
		}
		if entry.Type().IsRegular() {
			input, err := open(path)
			if err != nil {
				return err
			}
			_, copyErr := io.Copy(archive, input)
			closeErr := input.Close()
			return errors.Join(copyErr, closeErr)
		}
		return nil
	})
	return errors.Join(walkErr, archive.Close(), gz.Close(), file.Close())
}

type backupCreateOps struct {
	mkdir    func(string, os.FileMode) error
	archive  func(string, string, string) error
	snapshot func(context.Context, string, string) error
	stat     func(string) (os.FileInfo, error)
	hash     func(string) (string, error)
	revision func(repository.GitRepositoryPaths) (string, error)
	write    func(string, []byte, os.FileMode) error
}

func defaultBackupCreateOps() backupCreateOps {
	return backupCreateOps{os.MkdirAll, archiveDirectory, sqliteSnapshot, os.Stat, sha256File, repository.GetMainRevision, os.WriteFile}
}
func (r *Runtime) CreateBackup(ctx context.Context, destination string) (BackupManifest, error) {
	return r.createBackup(ctx, destination, defaultBackupCreateOps())
}
func (r *Runtime) createBackup(ctx context.Context, destination string, ops backupCreateOps) (BackupManifest, error) {
	if err := rejectNestedStatePath(destination, r.Paths.Root, "backup destination"); err != nil {
		return BackupManifest{}, err
	}
	if err := ops.mkdir(destination, 0777); err != nil {
		return BackupManifest{}, err
	}
	files := map[string]string{}
	repoArchive := filepath.Join(destination, "repo.git.tar.gz")
	if err := ops.archive(r.Paths.Repository.BareDir, repoArchive, "repo.git"); err != nil {
		return BackupManifest{}, err
	}
	controlCopy := filepath.Join(destination, "control.sqlite")
	if err := ops.snapshot(ctx, r.Paths.ControlDB, controlCopy); err != nil {
		return BackupManifest{}, err
	}
	names := []string{"repo.git.tar.gz", "control.sqlite"}
	if _, err := ops.stat(r.Paths.DerivedDB); err == nil {
		if err = ops.snapshot(ctx, r.Paths.DerivedDB, filepath.Join(destination, "derived.sqlite")); err != nil {
			return BackupManifest{}, err
		}
		names = append(names, "derived.sqlite")
	} else if !errors.Is(err, os.ErrNotExist) {
		return BackupManifest{}, err
	}
	sort.Strings(names)
	for _, name := range names {
		digest, err := ops.hash(filepath.Join(destination, name))
		if err != nil {
			return BackupManifest{}, err
		}
		files[name] = digest
	}
	revision, err := ops.revision(r.Paths.Repository)
	if err != nil {
		return BackupManifest{}, err
	}
	manifest := BackupManifest{1, revision, files}
	raw, _ := json.MarshalIndent(manifest, "", "  ")
	raw = append(raw, '\n')
	if err = ops.write(filepath.Join(destination, backupManifestName), raw, 0666); err != nil {
		return BackupManifest{}, err
	}
	return manifest, nil
}
func extractBackupArchive(source, destination string) error {
	return extractBackupArchiveWith(source, destination, os.MkdirAll, os.OpenFile)
}
func extractBackupArchiveWith(source, destination string, mkdir func(string, os.FileMode) error, openFile func(string, int, os.FileMode) (*os.File, error)) error {
	file, err := os.Open(source)
	if err != nil {
		return err
	}
	defer file.Close()
	gz, err := gzip.NewReader(file)
	if err != nil {
		return err
	}
	defer gz.Close()
	archive := tar.NewReader(gz)
	for {
		header, e := archive.Next()
		if errors.Is(e, io.EOF) {
			return nil
		}
		if e != nil {
			return e
		}
		if header.Typeflag == tar.TypeSymlink || header.Typeflag == tar.TypeLink {
			return fmt.Errorf("refusing to restore linked archive member: %s", header.Name)
		}
		clean := filepath.Clean(filepath.FromSlash(header.Name))
		if filepath.IsAbs(clean) || clean == ".." || strings.HasPrefix(clean, ".."+string(filepath.Separator)) {
			return fmt.Errorf("refusing to restore archive member outside destination: %s", header.Name)
		}
		target := filepath.Join(destination, clean)
		switch header.Typeflag {
		case tar.TypeDir:
			if e = mkdir(target, fs.FileMode(header.Mode)&0777); e != nil {
				return e
			}
		case tar.TypeReg:
			if e = mkdir(filepath.Dir(target), 0777); e != nil {
				return e
			}
			output, e := openFile(target, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, fs.FileMode(header.Mode)&0777)
			if e != nil {
				return e
			}
			_, copyErr := io.Copy(output, archive)
			closeErr := output.Close()
			if e = errors.Join(copyErr, closeErr); e != nil {
				return e
			}
		default:
			return fmt.Errorf("unsupported backup archive member: %s", header.Name)
		}
	}
}
func copyFile(source, target string) error { return copyFileWith(source, target, os.Open, os.OpenFile) }
func copyFileWith(source, target string, open func(string) (*os.File, error), openFile func(string, int, os.FileMode) (*os.File, error)) error {
	input, err := open(source)
	if err != nil {
		return err
	}
	defer input.Close()
	info, err := input.Stat()
	if err != nil {
		return err
	}
	output, err := os.OpenFile(target, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, info.Mode().Perm())
	if err != nil {
		return err
	}
	_, copyErr := io.Copy(output, input)
	return errors.Join(copyErr, output.Close())
}

type restoreBackupOps struct {
	lease       func(string, string) (*repository.WriterLease, error)
	manifest    func(string) (BackupManifest, error)
	stat        func(string) (os.FileInfo, error)
	temp        func(string, string) (string, error)
	mkdirAll    func(string, os.FileMode) error
	mkdir       func(string, os.FileMode) error
	removeAll   func(string) error
	extract     func(string, string) error
	copy        func(string, string) error
	revision    func(repository.GitRepositoryPaths) (string, error)
	materialize func(context.Context, repository.GitRepositoryPaths, string) (repository.MaterializedCheckout, error)
	rebuild     func(context.Context, *derived.Index, string, string) error
	readDir     func(string) ([]os.DirEntry, error)
	rename      func(string, string) error
}

func defaultRestoreBackupOps() restoreBackupOps {
	return restoreBackupOps{repository.AcquireWriterLease, LoadBackupManifest, os.Stat, os.MkdirTemp, os.MkdirAll, os.Mkdir, os.RemoveAll, extractBackupArchive, copyFile, repository.GetMainRevision, repository.MaterializeCurrentCheckout, func(ctx context.Context, index *derived.Index, root, revision string) error {
		return index.Rebuild(ctx, root, revision)
	}, os.ReadDir, os.Rename}
}
func RestoreBackup(ctx context.Context, config RuntimeConfig, directory string, rebuildDerived bool) (map[string]any, error) {
	return restoreBackup(ctx, config, directory, rebuildDerived, defaultRestoreBackupOps())
}
func restoreBackup(ctx context.Context, config RuntimeConfig, directory string, rebuildDerived bool, ops restoreBackupOps) (map[string]any, error) {
	paths := RuntimePathsFor(config)
	if err := rejectNestedStatePath(directory, paths.Root, "backup source"); err != nil {
		return nil, err
	}
	lease, err := ops.lease(paths.WriterLock, "memento-restore")
	if err != nil {
		return nil, err
	}
	defer lease.Release()
	manifest, err := ops.manifest(directory)
	if err != nil {
		return nil, err
	}
	parent := filepath.Dir(paths.Root)
	if _, e := ops.stat(parent); e != nil {
		parent = filepath.Dir(directory)
	}
	temp, err := ops.temp(parent, "memento-restore-")
	if err != nil {
		return nil, err
	}
	defer ops.removeAll(temp)
	staging := filepath.Join(temp, "state")
	if err = ops.mkdirAll(staging, 0777); err != nil {
		return nil, err
	}
	if err = ops.extract(filepath.Join(directory, "repo.git.tar.gz"), staging); err != nil {
		return nil, err
	}
	if err = ops.copy(filepath.Join(directory, "control.sqlite"), filepath.Join(staging, "control.sqlite")); err != nil {
		return nil, err
	}
	if !rebuildDerived && manifest.Files["derived.sqlite"] != "" {
		if err = ops.copy(filepath.Join(directory, "derived.sqlite"), filepath.Join(staging, "derived.sqlite")); err != nil {
			return nil, err
		}
	}
	staged := config
	staged.Repository.RootPath = staging
	stagedPaths := RuntimePathsFor(staged)
	revision, err := ops.revision(stagedPaths.Repository)
	if err != nil {
		return nil, err
	}
	if revision != manifest.RepoRevision {
		return nil, errors.New("backup manifest revision does not match archived main")
	}
	if _, err = ops.materialize(ctx, stagedPaths.Repository, revision); err != nil {
		return nil, err
	}
	if rebuildDerived {
		index := &derived.Index{Path: stagedPaths.DerivedDB}
		if err = ops.rebuild(ctx, index, stagedPaths.Repository.CurrentDir, revision); err != nil {
			return nil, err
		}
	}
	if err = ops.mkdirAll(paths.Root, 0777); err != nil {
		return nil, err
	}
	previous := filepath.Join(temp, "previous")
	if err = ops.mkdir(previous, 0777); err != nil {
		return nil, err
	}
	moved := []string{}
	installed := []string{}
	entries, err := ops.readDir(paths.Root)
	if err != nil {
		return nil, err
	}
	rollback := func() {
		for i := len(installed) - 1; i >= 0; i-- {
			name := installed[i]
			_ = ops.rename(filepath.Join(paths.Root, name), filepath.Join(staging, name))
		}
		for i := len(moved) - 1; i >= 0; i-- {
			name := moved[i]
			_ = ops.rename(filepath.Join(previous, name), filepath.Join(paths.Root, name))
		}
	}
	for _, entry := range entries {
		if entry.Name() == "locks" {
			continue
		}
		if err = ops.rename(filepath.Join(paths.Root, entry.Name()), filepath.Join(previous, entry.Name())); err != nil {
			rollback()
			return nil, err
		}
		moved = append(moved, entry.Name())
	}
	stagedEntries, err := ops.readDir(staging)
	if err != nil {
		rollback()
		return nil, err
	}
	for _, entry := range stagedEntries {
		if entry.Name() == "locks" {
			continue
		}
		if err = ops.rename(filepath.Join(staging, entry.Name()), filepath.Join(paths.Root, entry.Name())); err != nil {
			rollback()
			return nil, err
		}
		installed = append(installed, entry.Name())
	}
	return map[string]any{"repo_revision": manifest.RepoRevision, "restored_root": paths.Root, "rebuild_derived": rebuildDerived}, nil
}

type backupManifestOps struct {
	read func(string) ([]byte, error)
	stat func(string) (os.FileInfo, error)
	hash func(string) (string, error)
}

func LoadBackupManifest(directory string) (BackupManifest, error) {
	return loadBackupManifest(directory, backupManifestOps{os.ReadFile, os.Stat, sha256File})
}
func loadBackupManifest(directory string, ops backupManifestOps) (BackupManifest, error) {
	raw, err := ops.read(filepath.Join(directory, backupManifestName))
	if err != nil {
		return BackupManifest{}, err
	}
	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.DisallowUnknownFields()
	var manifest BackupManifest
	if err = decoder.Decode(&manifest); err != nil {
		return BackupManifest{}, err
	}
	if manifest.SchemaVersion != 1 || manifest.Files["repo.git.tar.gz"] == "" || manifest.Files["control.sqlite"] == "" {
		return BackupManifest{}, errors.New("backup manifest must include supported schema and required checksums")
	}
	for name, digest := range manifest.Files {
		if name != "repo.git.tar.gz" && name != "control.sqlite" && name != "derived.sqlite" {
			return BackupManifest{}, errors.New("unsupported backup manifest file")
		}
		path := filepath.Join(directory, name)
		if _, err = ops.stat(path); errors.Is(err, os.ErrNotExist) {
			return BackupManifest{}, fmt.Errorf("backup file is missing: %s", name)
		} else if err != nil {
			return BackupManifest{}, err
		}
		actual, e := ops.hash(path)
		if e != nil {
			return BackupManifest{}, e
		}
		if actual != digest {
			return BackupManifest{}, fmt.Errorf("checksum mismatch for %s", name)
		}
	}
	return manifest, nil
}
